### PR TITLE
std.child_process: enable non-standard streams

### PR DESCRIPTION
**Pull request by matu3ba**
This PR simplifies the setup code and handling to minimize leaking time
for both Windows and Posix and does not touch the list approach of
Windows. Portable pipes default to pipe2 with CLOEXEC on Posix and
disabled handle inhertence on Windows except shortly before
and after the `spawn`.

The potential time for leaking can be long due trying to spawn on
multiple PATH and PATHEXT entries on Windows.

Leaking on Posix
1. CLOEXEC does not take immediately effect with clone(), but after the
setup code for the child process was run and a exec* function is executed.
External code may at worst be long living and never do exec*.
2. File descriptors without CLOEXEC are designed to "leak to the child",
but every spawned process at the same time gets them as well.

File leaking on Windows can be fixed with an explicit list approach,
which might require runtime probing to not break on Kernel updates
for the parameter list.

### CODE REVIEW

**Comment by matu3ba** [comment]
Issue left on windows before cleanup:
```
thread 5652 panic: reached unreachable code
D:\a\zig\zig\build-release\stage3-release\lib\zig\std\debug.zig:285:14: 0x7ff72eb64b47 in assert (parent.exe.obj)
    if (!ok) unreachable; // assertion failure
             ^
D:\a\zig\zig\test\standalone\childprocess_extrapipe\parent.zig:92:48: 0x7ff72eb64278 in main (parent.exe.obj)
        std.debug.assert(handle_flags & windows.HANDLE_FLAG_INHERIT != 0);
                                               ^
```

File: test/standalone/childprocess_extrapipe/parent.zig [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    );
+    {
+        // close read side of pipe
+        if (comptime builtin.target.isDarwin()) {
+            {
+                // less time for leaking the file descriptor, if it is closed immediately
+                child_proc.posix_actions = try os.posix_spawn.Actions.init();
+                errdefer child_proc.posix_actions.?.deinit();
+                try child_proc.posix_actions.?.close(pipe[1]);
+                errdefer child_proc.posix_actions.?.deinit();
```
Comment (matu3ba) [inline test/standalone/childprocess_extrapipe/parent.zig:68]: I did forget to remove this one.

File: test/standalone/childprocess_extrapipe/parent.zig [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    else
+        try handleToString(pipe[0], &buf);
+
+    var child_proc = ChildProcess.init(
+        &.{ child_path, s_handle },
+        gpa,
+    );
+    {
+        // close read side of pipe
+        if (comptime builtin.target.isDarwin()) {
```
Comment (matu3ba) [inline test/standalone/childprocess_extrapipe/parent.zig:62]: Hey @kubkon,
I'd like to have a quick feedback before removing the clunkyness by pipe numbers and the used optionals of Windows.

This block feels very meh, especially the `errdefer` statement, which is a footgun. However, I dont know any other way to keep the defaults of "simple init and spawn" and not force the handling to be always done by the user.
This code is more of an example how user-handling of posix_spawn together with keeping the default would looks like and not strictly necessary.

Any opinions what I should do with the posix_spawn stuff, @kubkon ?

File: lib/std/child_process.zig [hunk 3]
```diff
@@ -57,6 +57,12 @@ pub const ChildProcess = struct {
 
     expand_arg0: Arg0Expand,
 
+    /// Use this for additional posix spawn attributes.
```
Comment (matu3ba) [inline lib/std/child_process.zig:60]: These are the other relevant bits to know for users.

File: test/standalone/childprocess_extrapipe/parent.zig (hunk 2 above)
Comment (matu3ba) [reply test/standalone/childprocess_extrapipe/parent.zig:62]: I think a temporary to keep track in the spawn() method, if the user has set `actions` and `attributes` and not doing `deinit()` sounds more reasonable than "partial deinit on error case".

**Comment by matu3ba** [comment]
I should always cross-compile locally:
```
 FAILED: zig2.c D:/a/zig/zig/build-release/zig2.c 
cmd.exe /C "cd /D D:\a\zig\zig && D:\a\zig\zig\build-release\zig1.exe D:/a/zig/zig/lib build-exe src/main.zig -ofmt=c -lc -OReleaseSmall --name zig2 -femit-bin="D:/a/zig/zig/build-release/zig2.c" --pkg-begin build_options D:/a/zig/zig/build-release/config.zig --pkg-end -target x86_64-windows"
/lib/std/child_process.zig:899:21: error: expected type '*?*anyopaque', found '**anyopaque'
                    &tmp_hChildStd_Rd,
                    ^~~~~~~~~~~~~~~~~
/lib/std/child_process.zig:899:21: note: pointer type child '*anyopaque' cannot cast into pointer type child '?*anyopaque'
/lib/std/child_process.zig:899:21: note: mutable '*anyopaque' allows illegal null values stored to type '?*anyopaque'
/lib/std/child_process.zig:1510:9: note: parameter type declared here
    rd: *?windows.HANDLE,
        ^~~~~~~~~~~~~~~~
referenced by:
    spawn: /lib/std/child_process.zig:157:24
    linkWithLLD: src/link/Wasm.zig:3604:26
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```
see
```
/home/ci/actions-runner/_work/zig/zig/lib/std/child_process.zig:899:21: error: expected type '*?*anyopaque', found '**anyopaque'
                    &tmp_hChildStd_Rd,
                    ^~~~~~~~~~~~~~~~~
/home/ci/actions-runner/_work/zig/zig/lib/std/child_process.zig:899:21: note: pointer type child '*anyopaque' cannot cast into pointer type child '?*anyopaque'
/home/ci/actions-runner/_work/zig/zig/lib/std/child_process.zig:899:21: note: mutable '*anyopaque' allows illegal null values stored to type '?*anyopaque'
/home/ci/actions-runner/_work/zig/zig/lib/std/child_process.zig:1510:9: note: parameter type declared here
    rd: *?windows.HANDLE,
        ^~~~~~~~~~~~~~~~
referenced by:
    spawn: /home/ci/actions-runner/_work/zig/zig/lib/std/child_process.zig:157:24
    execAllowFail: /home/ci/actions-runner/_work/zig/zig/lib/std/build.zig:1197:18
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

error: test...
error: The following command exited with error code 1:
/home/ci/actions-runner/_work/zig/zig/build-release/stage3-release/bin/zig test /home/ci/actions-runner/_work/zig/zig/lib/std/std.zig -lc --test-name-prefix std-aarch64-windows-gnu-Debug-c-multi-default  --cache-dir /home/ci/actions-runner/_work/zig/zig/build-release/zig-local-cache --global-cache-dir /home/ci/actions-runner/_work/zig/zig/build-release/zig-global-cache --name test -fno-single-threaded -target aarch64-windows-gnu -mcpu generic --test-no-exec -I /home/ci/actions-runner/_work/zig/zig/test -L /home/ci/deps/zig+llvm+lld+clang-x86_64-linux-musl-0.11.0-dev.971+19056cb68/lib -I /home/ci/deps/zig+llvm+lld+clang-x86_64-linux-musl-0.11.0-dev.971+19056cb68/include --zig-lib-dir /home/ci/actions-runner/_work/zig/zig/lib --enable-cache 
```

File: lib/std/child_process.zig (hunk 3 above)
Comment (matu3ba) [reply lib/std/child_process.zig:60]: removed with different api: either user takes ownership and must deinit or (default) its within the macos posix_spawn function.

**Comment by matu3ba** [comment]
There is no simple way to unify declarative and imperative approaches of posix_spawn and non-posix-spawn, because the former allocates and the latter not.

- [x] improve phrasing etc of this pr description
- [ ] Provide nicer abstraction than pipe[0], pipe[1] for read and write end of pipe.
- [ ] portability wrapper function: open pipe with CLOEXEC on default or fcntl on it with the CLOEXEC always (better slow and correct than leaking and have bugs)
- [ ] provide a way to handle a list of descriptors the child process should inherit both on Unixes and Windows (CLOEXEC removed before fork/clone call and set again on other side)
- [ ] accept, that it cant be made very user friendly:
- [ ] 1. file handle not to be used in parent needs manual cleanup after run call to prevent more clunky allocation
- [ ] 2. windows needs disabling of file handle inheritance of the inherited handle in child process, need to recheck if necessary on fork()

**Comment by matu3ba** [comment]
known to work commit `fix windows async pipe creation`

File: lib/std/os.zig [hunk 4]
```diff
@@ -44,6 +44,32 @@ pub const windows = @import("os/windows.zig");
 pub const posix_spawn = @import("os/posix_spawn.zig");
 pub const ptrace = @import("os/ptrace.zig");
 
+/// Pipe read side
```
Comment (matu3ba) [inline lib/std/os.zig:47]: Move into child_process. A namespace `Pipe` sounds nice, but I don't know things to move inside except for init and the constants.

File: lib/std/os.zig [hunk 5]
```diff
@@ -44,6 +44,32 @@ pub const windows = @import("os/windows.zig");
...
+pub const pipe_rd = 0;
+/// Pipe write side
+pub const pipe_wr = 1;
+
+pub const windowsPtrDigits: usize = std.math.log10(math.maxInt(usize));
+pub const otherPtrDigits: usize = std.math.log10(math.maxInt(u32)) + 1; // +1 for sign
+pub const handleCharSize = if (builtin.target.os.tag == .windows) windowsPtrDigits else otherPtrDigits;
+
+pub fn handleToString(handle: fd_t, buf: []u8) std.fmt.BufPrintError![]u8 {
```
Comment (matu3ba) [inline lib/std/os.zig:56]: name suggestions other than handleToString? Its called handle on Windows and file descriptor on Unixoids.
For now I used the shorter name.

File: lib/std/os.zig [hunk 6]
```diff
@@ -44,6 +44,32 @@ pub const windows = @import("os/windows.zig");
 pub const posix_spawn = @import("os/posix_spawn.zig");
 pub const ptrace = @import("os/ptrace.zig");
 
+/// Pipe read side
+pub const pipe_rd = 0;
+/// Pipe write side
+pub const pipe_wr = 1;
+
+pub const windowsPtrDigits: usize = std.math.log10(math.maxInt(usize));
```
Comment (matu3ba) [inline lib/std/os.zig:52]: These should be computed numbers.

File: lib/std/os.zig [hunk 7]
```diff
@@ -4843,6 +4869,28 @@ pub fn lseek_CUR_get(fd: fd_t) SeekError!u64 {
     }
 }
 
+const EnableInheritanceError = FcntlError || windows.SetHandleInformationError;
```
Comment (matu3ba) [inline lib/std/os.zig:4872]: This is missing the portable probing info if inheritance is enabled, which sounds relevant for developer.

File: test/standalone/childprocess_extrapipe/child.zig [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    defer if (general_purpose_allocator.deinit()) @panic("found memory leaks");
+    const gpa = general_purpose_allocator.allocator();
+
+    var it = try std.process.argsWithAllocator(gpa);
+    defer it.deinit();
+    _ = it.next() orelse unreachable; // skip binary name
+    const s_handle = it.next() orelse unreachable;
+    var file_handle = try std.os.stringToHandle(s_handle);
+
+    if (builtin.target.os.tag == .windows) {
```
Comment (matu3ba) [inline test/standalone/childprocess_extrapipe/child.zig:15]: portable handle probing here.

File: test/standalone/childprocess_extrapipe/parent.zig [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    // less time to leak => better
+    {
+        try os.enableInheritance(pipe[pipe_rd]);
+        defer os.close(pipe[pipe_rd]);
+
+        try child_proc.spawn();
+    }
+
+    // check that inheritance was disabled for the handle the whole time
+    if (builtin.target.os.tag == .windows) {
```
Comment (matu3ba) [inline test/standalone/childprocess_extrapipe/parent.zig:42]: also inheritance probing.

**Review by matu3ba — COMMENTED** [review]
todos.

**Comment by matu3ba** [comment]
working build: `fixup build error.`

File: lib/std/os/windows.zig [hunk 10]
```diff
@@ -1566,6 +1576,45 @@ pub fn GetEnvironmentVariableW(lpName: LPWSTR, lpBuffer: [*]u16, nSize: DWORD) G
     return rc;
 }
 
+// zig fmt: off
```
Comment (matu3ba) [inline lib/std/os/windows.zig:1579]: I kept these inside for https://github.com/ziglang/zig/issues/14251, but they are generally useful for other users as well short of advanced list attributes.

**Comment by matu3ba** [comment]
I feel like this is ready for review once CI is successful. While this is short of #14251 to prevent all leaks on Windows, I think this is very nice to use.

**Comment by andrewrk** [comment]
It is unclear to me whether this was ever ready for review. Either way, it has succumbed to bitrot and must be refreshed in a new PR against the latest master. If you do refresh this in a new PR, please try to make it easier to review, clearly explaining the problem you are solving with a motivating use case, and don't leave open questions such as a comment on your own code that says "I dont like this solution.". Instead, do your best and then be clear that you want it to be merged.

**Pull request closed without merging**